### PR TITLE
fix(downloads): let a failed download reach failed-handling from InterDir

### LINKED ISSUE #805: Failed download never reaches failed-handling when the payload sits outside a post-processing root
## Summary

A failed download never reaches failed-download handling when the download client leaves the payload outside a configured post-processing root. The post-process command is rejected first, the release parks in `manual_review`, the failure is never recorded, and **the identical release is re-grabbed on every subsequent search, indefinitely**.

On my install this had seven issues of one series cycling through the same release six times each, and `"Initiating Failed Download handling"` appears **zero times in the entire log**.

## Root cause

`validate_postprocess_item()` validates `nzb_folder` containment unconditionally, before anything inspects the `failed` flag:

```python
# comicarr/app/downloads/pp_commands.py
command["nzb_folder"] = str(
    validate_mapped_path(command.get("nzb_folder"), roots=roots, require_exists=require_exists)
)
```

For a **failed** download, `nzbget.py` reports the folder as the parent of `DestDir`:

```python
"location": os.path.abspath(os.path.join(hq[0]["DestDir"], os.pardir)),
```

NZBGet leaves a `FAILURE/*` item under `InterDir`, so that resolves to `InterDir`'s parent — not the completed-download root. Containment fails:

```
[DOWNLOADS-PP] Rejected unsafe post-processing command:
    nzb_folder is outside configured post-processing roots
```

`PostProcessCommandError` propagates, the item is quarantined, and `process.py` — which routes `failed is True` to `FailedProcessor` — is never reached.

## Why it becomes permanent

`manual_review` is terminal (`retry_count` and `next_retry_at` both NULL), so the row never self-heals. And because `FailedProcessor.markFailed()` never runs, the `failed` table stays empty — so `FAILED_DOWNLOAD_CHECKER` in `search.py` never trips and never skips the bad release. Every later search re-selects it, re-fails identically, and re-parks.

Observed on a live install before the fix:

```
failed table rows:                                 0
"Initiating Failed Download handling" in log:      0
"nzb_folder is outside configured post-processing roots": every attempt
pipeline_journal: 7 rows in manual_review/manual_review, retry_count 5-6
```

## Reproduce

1. NZBGet with `DupeCheck=yes` (or any path to a `FAILURE/*` status), `InterDir` outside `nzbget_directory`
2. Let Comicarr grab a release that NZBGet fails (a duplicate grab produces `FAILURE/BAD`)
3. The item parks in `manual_review` with `invalid_postprocess_command`
4. "Search again" re-selects the same release and repeats forever

## Proposed fix

A failed download is never imported from that folder — it routes to failed-download handling, which only needs to know the release failed so it can look for a different one. Sanitize the path, but don't require containment:

```python
if _is_failed_download(command.get("failed")):
    command["nzb_folder"] = str(_canonical_path(command.get("nzb_folder"), "nzb_folder"))
else:
    command["nzb_folder"] = str(validate_mapped_path(...))
```

Successful downloads keep the containment boundary unchanged. Traversal/separator sanitising still applies to both.

## Verified

After applying the fix to `v0.38.2` and deploying:

| | before | after |
|---|---|---|
| `nzb_folder is outside configured post-processing roots` | every attempt | 0 |
| `Initiating Failed Download handling` | 0 (all history) | runs |
| `failed` table | 0 rows | records the release |
| `manual_review/manual_review` rows | 7 | 0 |
| Needs-attention panel | 7 issues | empty |

Full unit suite passes (2814). PR follows with a regression test that fails on `main` with the exact production error.

### PR DESCRIPTION
**Pull request by azuretek**
Fixes #805.

Two fixes an operator will notice, both on the path from a grab to a usable file. They ride together because the second is what makes the first recoverable: failed-handling is only useful if the follow-up search can actually find a different release.

Covered by one changeset, `.changeset/failed-download-handling-and-search-alternatives.md`.

---

## 1. A failed download never reached failed-handling

`validate_postprocess_item()` requires `nzb_folder` to sit under a configured post-processing root **before** anything inspects the `failed` flag. But a failed download's folder legitimately sits outside those roots: NZBGet leaves a `FAILURE/*` item under `InterDir`, and `nzbget.py` reports the folder as `DestDir`'s parent.

So the command is rejected:

```
[DOWNLOADS-PP] Rejected unsafe post-processing command:
    nzb_folder is outside configured post-processing roots
```

`PostProcessCommandError` propagates and the item is quarantined, so `process.py` — which routes `failed is True` to `FailedProcessor` — is never reached. Because `markFailed()` never runs, the `failed` table stays empty, `FAILED_DOWNLOAD_CHECKER` never trips, and the same release is re-grabbed on every later search. `manual_review` is terminal, so the row never self-heals.

**Change.** A failed download is never imported from that folder — it only needs to reach failed-handling so a different release can be found. Sanitize the path, but don't require containment.

Successful downloads are untouched and still confined to the configured roots. Traversal and separator sanitising still applies to both paths.

The `failed` flag is also now type-aware rather than a membership test, and is normalized to a bool at the command boundary. `bool` subclasses `int` and numeric types compare across themselves, so `1.0 == 1 == True` let a float take the relaxed branch; integer `1` was accepted at the boundary but then missed `Process.post_process()`'s identity check, which only rewrites the string forms — leaving the item stuck. `False` / `"0"` / `None` still get full containment validation.

## 2. One failed release rejected every alternative (`comicarr/search.py`)

`verification()` passed the whole `verified_matches` list to `searcher()`, which reads `ComicName`, `nzbid`, size and title off `comicinfo[0]`. So every candidate was described by candidate 0: the failed-release check ran against candidate 0's nzbid no matter which candidate was in hand, and once one release for an issue had failed, every alternative the provider offered was rejected too. The issue stayed Wanted with nothing to show for the search — the exact state fix 1 leaves an issue in.

Each candidate is now handed to `searcher()` as itself.

Making fallbacks reachable exposed a second defect in the same function, [raised in review](https://github.com/frankieramirez/comicarr/pull/807#discussion_r3909402591): `verified_index` only advanced inside the `downloadit` branch, but `search_filer` emits `alt_match` entries with `downloadit=False` into the same list. A skipped entry ahead of the accepted candidate left the index pointing at the earlier row, and the post-loop block reads `verified_matches[verified_index]` for the pack check, the nzbid it hands `nzblog()`, and the snatch year — so the accepted release was recorded under the wrong nzbid and post-processing could not match the download back to the issue. The index now comes from `enumerate()`. `nzbname_create()` reads `info[0]` for the blackhole name, the same shape as `searcher()`, so it gets the current candidate too.

> This is the same change as #807, cherry-picked here so the two fixes could be verified together on a live install. Happy to drop it from this branch and let #807 land on its own if you'd rather keep them separate — just say which you prefer.

## Tests

`test_failed_download_is_not_required_to_sit_under_a_postprocess_root` covers:

- a failed download from outside the roots is accepted
- both explicit truthy forms (`True`, `"1"`)
- `False` / `"0"` / `None` still get full containment validation
- traversal is still rejected for a failed download
- `1.0` and `0.0` are rejected; integer `1` is accepted and normalized

`test_search_verification_candidate.py` covers:

- each candidate is described by itself, not by candidate 0
- a failed first candidate does not stop the second from being sent
- a leading `downloadit=False` entry does not shift the nzbid handed to `nzblog()`

On `main` the first fails with the exact production error:

```
E   comicarr.app.downloads.pp_commands.PostProcessCommandError:
    nzb_folder is outside configured post-processing roots
```

and the indexing test fails with `nzblog was given the skipped candidate's nzbid: 'alt-id'`.

Full unit suite passes: **2817 passed**. `ruff check`, `ruff format --check` and `lint:guards` clean.

## Verified on a live install

Applied to `v0.38.2`, built and deployed against the reproducing install:

| | before | after |
|---|---|---|
| `nzb_folder is outside configured post-processing roots` | every attempt | 0 |
| `Initiating Failed Download handling` | 0 in all history | runs |
| `failed` table | 0 rows | records the release |
| `pipeline_journal` `manual_review/manual_review` | 7 | 0 |
| Needs-attention panel | 7 issues | empty |

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved handling of failed downloads whose folders are located outside configured post-processing directories.
  - Preserved security checks for non-failed downloads and path traversal attempts.
  - Download failure status is now recognized only for explicit failure values.

### CODE REVIEW

File: comicarr/app/downloads/pp_commands.py [hunk 1]
```diff
@@ -85,15 +85,34 @@ def _validate_basename(value):
     return name
 
 
+def _is_failed_download(value):
+    """True only for an explicit failed flag; anything else gets full validation."""
+    return value in (True, 1, "1")
```
[the remark this answers is not in the document]
Comment (azuretek) [reply comicarr/app/downloads/pp_commands.py]: Fixed in d86d4a2 — and the merge-risk note turned out to be the more important half, so I addressed both.

**`1.0` (the inline finding).** `_is_failed_download` and the `failed`/`apicall`/`ddl`/`oneoff` flag check were both membership tests. `bool` subclasses `int` and numeric types compare across themselves, so `1.0 == 1 == True` and a float took the relaxed path; the flag check additionally accepted `0.0`. Both are now `isinstance` based, with `bool` tested first because `isinstance(True, int)` is also true.

I also moved the flag check ahead of the folder branch, so an unsupported form is reported as `<key> must be boolean` rather than surfacing as a confusing path rejection.

**Integer `1` (the merge-risk note).** This was a real stuck state, not just an inconsistency. `Process.post_process()` rewrites only the string forms `"0"`/`"1"` and then branches on `failed is False` / `failed is True` — identity, not equality — so an integer `1` matches neither: no `PostProcessor`, no `FailedProcessor`, item silently does nothing. That is the same dead end this PR is fixing, reached by a different route. Tightening the membership test alone would not have caught it, since `1` is a legitimately accepted form.

So `failed` is now canonicalized to a real bool at the boundary where the command is already proven safe. Every accepted truthy form reaches failed-download handling; every falsey form (including a present-but-`None` flag, which previously also matched neither branch) reaches the import path. Accepted values are unchanged: `True`/`1`/`"1"` are failed, `False`/`0`/`"0"`/`None` are not.

**Coverage.** `test_failed_download_is_not_required_to_sit_under_a_postprocess_root` now asserts `command["failed"] is True` for each of `True`/`1`/`"1"`, `is False` for each of `False`/`0`/`"0"`/`None`, and rejection of `1.0`/`0.0`/`2`/`"true"`/`"yes"` for all four flags with the `must be boolean` message.

Verified A/B against the pre-fix validator rather than by absence of errors — the new assertions fail on stock with `assert 1 is True`, and the old predicates accept `1.0` where the new ones reject it:

```
   value | old_failed old_bool | new_failed new_bool
    True |       True     True |       True    True
       1 |       True     True |       True    True
     '1' |       True     True |       True    True
     1.0 |       True     True |      False   False
     0.0 |      False     True |      False   False
    None |      False     True |      False    True
```

Full unit suite green (2814 passed), `ruff check` and `ruff format --check` clean.

**Review by frankieramirez — COMMENTED** [review]
This needs a changeset before it can merge. Both fixes are things an operator will notice (failed downloads stop parking in Needs Attention, and search stops throwing away every alternative after one failed release), so a single patch entry covering both is enough.

The search.py change also isn't in the description. Could you add a line about it there?

**Comment by azuretek** [comment]
No problem — both done, check it out now!

**Changeset:** added as `.changeset/failed-download-handling-and-search-alternatives.md`, one `patch` entry covering both, written outcome-first per CONTRIBUTING:

> Failed downloads now reach failed-handling instead of parking in Needs Attention. A download NZBGet reported as failed sits in `InterDir`, outside the configured post-processing roots, so the command was rejected on its path before anything looked at the failed flag. The grab was never marked failed, the failed-download checker never tripped, and the same broken release was re-grabbed on every later search while the issue stayed stuck in Needs Attention with no way to clear it. Failed downloads are still path-sanitised, and successful ones are still confined to the configured roots.
>
> A previously failed release no longer rejects every alternative for the same issue. Every candidate the provider returned was checked against the *first* candidate's release id, so one failed release discarded the whole result set and the issue stayed Wanted with nothing to show for the search. Each candidate is now judged on its own, and the release that is actually snatched is the one recorded against the issue — so post-processing can match the download back to it.

**Description:** rewritten so the search change is a named section rather than an unexplained file in the diff. I also framed why they're together, since it's the honest reason they were tested as a pair — failed-handling only helps if the follow-up search can actually reach a different release, and before this it couldn't.

**One thing worth your call:** that search commit is the same change as #807, cherry-picked here so both could be verified on the same live install. So these two PRs overlap, and whichever lands first empties the other's `search.py` diff. Happy either way — say the word and I'll drop it from this branch and let #807 stand alone, or close #807 and let this carry both.

Also carried over the fix you raised on #807 (`verified_index` not advancing past `downloadit=False` alt_match entries) so this branch doesn't merge the version without it.

Full suite green here: **2817 passed**, `ruff check` / `ruff format --check` / `lint:guards` clean.

**Comment by frankieramirez** [comment]
Thanks, this looks good. Keep the search change here since the changeset already covers both and they were verified together. Go ahead and close #807 so we don't end up with two copies of the same diff.

**Review by frankieramirez — APPROVED** [review]
